buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java

```diff
@@ -542,6 +539,25 @@ public void testIsTextWithInvalidIndexAndLength() {
         }
     }
 
+    @Test
+    public void testUtf8BytesIsExact() {
+        final String utf8 = "Some UTF-8 like äÄ∏ŒŒ";
+        checkUtf8Bytes(utf8);
+    }
+
+    @Test
+    public void testUtf8BytesWithSurrogatesIsExact() {
+        final String utf8 = "a\uD800\uDC00b";
+        checkUtf8Bytes(utf8);
+    }
+
+    private static void checkUtf8Bytes(final CharSequence utf8) {
+        final ByteBuf buf = Unpooled.buffer(ByteBufUtil.utf8MaxBytes(utf8));
+        final int writtenBytes = ByteBufUtil.writeUtf8(buf, utf8);
```

Comment:
It worries me to see this same class used as the testing oracle. Could this be `utf8.toString().getBytes(CharsetUtil.UTF_8).length` or similar?

Reply:
Makes sense, let me check what I can do!

Reply:
I've looked into it and probably it isn't the case here.
I'm not assuming that `ByteBufUtil::writeUtf8` and `String::getBytes` for UTF-8 charset will produce 1:1 the same output (eg just consider the `?` replacement done on `ByteBufUtil::writeUtf8`), hence I prefer to have these methods coupled as the `checUtf8Bytes` javadoc states.

Reply:
I guess that's fine because the return value will be too small or there will be an index out of range exception. Although there's not any test that triggers WRITE_UTF_UNKNOWN.